src/bundle.ts

```diff
@@ -2,49 +2,161 @@ import * as os from "os";
+	// The `file` is a dummy, meant to allow Karma to work. But, we can't write
+	// to it without causing Karma to refresh. So, we have a real file that we
+	// write to, and allow esbuild to build from.
 	file = path.join(this.dir, `${random(16)}-bundle.js`);
+	private realFile = this.file.replace(/\.js$/, ".mjs");
+
+	// Dirty signifies that new data has been written, and is cleared once a build starts.
+	private _dirty = false;
+	// buildsInProgress tracks the number of builds. When a build takes too long, a new build
+	// may have started before the original completed. In this case, we resolve the build with
+	// the latest result.
+	private buildsInProgress = 0;
+	private deferred = new Deferred<BundledFile>();
+	private incrementalBuild: esbuild.BuildIncremental | null = null;
+
+	constructor(log: Logger, config: esbuild.BuildOptions) {
+		this.log = log;
+		this.config = config;
+	}
 
 	addFile(file: string) {
 		const normalized = path
-			.relative(this.dir, file)
+			.relative(this.file, file)
 			.replace(/\\/g, path.posix.sep);
 		this.files.add(normalized);
 	}
 
-	generate() {
-		const files = Array.from(this.files).map(file => {
-			return `import "${file}";`;
-		});
-		return files.join("\n");
+	dirty() {
+		if (this._dirty) return;
+		this._dirty = true;
+		this.deferred = new Deferred();
 	}
 
-	write(item: BundledFile) {
-		this.item = item;
+	async write(beforeProcess: () => {}, afterProcess: () => {}) {
+		if (this.buildsInProgress === 0) beforeProcess();
+		this.buildsInProgress++;
+
+		const { deferred } = this;
+		const result = await this.bundle();
+
+		// The build took so long, we've already had another write.
+		// Instead of serving a stale build, let's wait for the new one to resolve.
+		this.buildsInProgress--;
+		if (this.buildsInProgress > 0 || this._dirty) {
+			deferred.resolve(this.deferred.promise);
+			return deferred.promise;
+		}
+
+		afterProcess();
+		deferred.resolve(result);
+		return result;
 	}
 
 	read() {
-		return this.item;
+		return this.deferred.promise;
 	}
 
 	touch() {
 		fs.writeFileSync(this.file, "");
+		fs.writeFileSync(this.realFile, "");
+	}
+
+	async stop() {
+		// Wait for any in-progress builds to finish. At this point, we know no
+		// new ones will come in, we're just waiting for the current one to
+		// finish running.
+		if (this.buildsInProgress > 0 || this._dirty) {
+			await this.deferred.promise;
+		}
+		// Releasing the result allows the child process to end.
+		this.incrementalBuild?.rebuild.dispose();
+		this.incrementalBuild = null;
+	}
+
+	private async bundle() {
+		try {
+			if (this._dirty) {
+				this._dirty = false;
+				const files = Array.from(this.files).map(file => {
+					return `import "${file}";`;
+				});
+				fs.writeFileSync(this.realFile, files.join("\n"));
+			}
+			if (this.incrementalBuild) {
+				const result = await this.incrementalBuild.rebuild();
+				return this.processResult(result as BuildResult);
+			}
+
+			const result = (await esbuild.build({
```

Comment:
This is great! Solves in #19 in one go as the service API is no more starting with esbuild `0.9.0` 👍